Adding telemetry page (#8392)
fishtown-analytics/docs.getdbt.com: website/docs/docs/fusion/telemetry-observability.md

## Node outcome

Every node produces a result for each phase it participates in. Some phases, such as `parse`, don't involve node-level execution, so they don't produce node spans or node outcomes.

 The `node_outcome` field indicates whether or not <Constant name="fusion" /> executed the node's operation. 

| Outcome    | Description                                                    |
| ---------- | -------------------------------------------------------------- |
| `success`  | The node operation was completed with no errors.   |
| `error`    | The node operation failed to execute (for example, syntax error). |
| `skipped`  | The node was not evaluated (see [skip reasons](#skip-reasons)). |
| `canceled` | The node was interrupted (for example, user pressed Ctrl+C).  |

### Skip reasons

When <Constant name="fusion" /> skips a node, the telemetry includes a reason:

| Skip reason      | Description                                                      |
| ---------------- | ---------------------------------------------------------------- |
| `upstream`       | A dependency failed.                                              |

> | `cached`         | <Constant name="fusion" /> reused results from cache (no changes detected). |

Author: ```suggestion
| `cached`         | <Constant name="fusion" /> reused results from cache (no changes detected via [state aware orchestration](/docs/deploy/state-aware-about)). |
```

---
title: "Telemetry and observability"
id: "telemetry"
description: "Fusion support for telemetry and observability"
pagination_next: null
pagination_prev: null
---

The <Constant name="fusion_engine" /> provides a comprehensive telemetry system that replaces [<Constant name="core" />'s structured logging](/reference/events-logging#structured-logging). Built on [OpenTelemetry](https://opentelemetry.io/) conventions and backed by a stable protobuf schema, it enables deep integration with orchestrators, observability platforms, and custom tooling.

This uses the same integration that <Constant name="dbt_platform" /> relies on for orchestration and monitoring, providing proven and production-ready features that work at scale.

## Available output formats

<Constant name="fusion" /> telemetry supports three output formats, which you can enable independently:

> | Format      | Use case                                                         | Availability               |

Reviewer: this is a great table! however i'd love to have more info how which format works best with my tooling?

sugg adding a 'works well with' or 'typical integrations' column (like you did wiht OTLP)

Author: Will look to expand on this in the next iteration

| ----------- | ---------------------------------------------------------------- | -------------------------- |

> | **JSONL**   | Real-time monitoring, streaming to downstream systems            | Written as events occur    |

Reviewer: ```suggestion
| **JSONL**   | Real-time monitoring, streaming to downstream systems.            | Written as events occur.    |
```

> | **Parquet** | Post-run analysis, querying, and long-term storage               | Written when runs complete |

Reviewer: ```suggestion
| **Parquet** | Post-run analysis, querying, and long-term storage.               | Written when runs complete. |
```

> | **OTLP**    | Integration with observability platforms (Datadog, Jaeger, and more) | Streamed in real-time  |

Reviewer: ```suggestion
| **OTLP**    | Integration with observability platforms (Datadog, Jaeger, and more). | Streamed in real-time.  |
```

Write JSONL to a file (saves to the `logs/` directory):

```bash
dbtf build --otel-file-name telemetry.jsonl
```

Stream JSONL to stdout:
```bash
dbtf build --log-format otel
```

Write a Parquet file (saves to `target/metadata/` directory):
```bash
dbtf build --otel-parquet-file-name telemetry.parquet
```

Export to an OpenTelemetry collector:
```bash
OTEL_EXPORTER_OTLP_ENDPOINT="http://localhost:4318" dbtf build --export-to-otlp
```

## Telemetry data

<Constant name="fusion" /> telemetry contains two types of records:

- **Spans** &mdash; Operations with a start and end time (like compiling a model or running a test).
- **Log records** &mdash; Point-in-time events within a span.

### Telemetry hierarchy

Every dbt command creates a hierarchy of spans:

```
Invocation (dbtf build)
├── Phase (Parse)
├── Phase (Compile)
│   ├── Node (model.project.customers)
│   └── Node (model.project.orders)
└── Phase (Run)
    ├── Node (model.project.customers)
    └── Node (model.project.orders)
```

The `trace_id` (also known as `invocation_id`) remains consistent across all telemetry records for a single dbt command, making it easy to correlate events.

## Node outcome

> Every node produces a result for each phase it participates in. The `node_outcome` field indicates whether or not <Constant name="fusion" /> executed the node's operation. 

Reviewer: i had to read this a couple of times to catch the 'it participates in'. we shouldn't have readers infer behavior from a conditional clause so sugg making it clearer

```suggestion
Every node produces a result for each phase it participates in. Some phases, such as `parse`, don't involve node-level execution, so they don't produce node spans or node outcomes.

 The `node_outcome` field indicates whether or not <Constant name="fusion" /> executed the node's operation. 
```

Reviewer: or do we want to say `parse`?

| Outcome    | Description                                                    |
| ---------- | -------------------------------------------------------------- |
| `success`  | The node operation was completed with no errors.   |
| `error`    | The node operation failed to execute (for example, syntax error). |
| `skipped`  | The node was not evaluated (see [skip reasons](#skip-reasons)). |
| `canceled` | The node was interrupted (for example, user pressed Ctrl+C).  |

### Skip reasons

When <Constant name="fusion" /> skips a node, the telemetry includes a reason:

| Skip reason      | Description                                                      |
| ---------------- | ---------------------------------------------------------------- |
| `upstream`       | A dependency failed.                                              |

> | `cached`         | <Constant name="fusion" /> reused results from cache (no changes detected). |

Reviewer 2: we should link this to SAO page, our flagship feature! Sorry, I forgot to mention that in my draft

| `phase_disabled` | The phase was disabled (for example, `--static-analysis off`).    |
| `noop`           | Node doesn't perform work in this phase (for example, ephemeral models). |

### Test outcomes

When a test executes successfully (`node_outcome: success`), it reports the test result:

| Test outcome | Description                                   |
| ------------ | --------------------------------------------- |

> | `passed`     | No failures detected .                         |

Reviewer: ```suggestion
| `passed`     | No failures detected.                         |
```

| `warned`     | Failures detected, but configured as warnings. |
| `failed`     | Failures detected (data quality issue).        |

:::info Test outcomes

A test with `node_outcome: success` and `test_outcome: failed` means <Constant name="fusion" /> successfully ran the test, and the test reported data quality issues. This differs from `node_outcome: error`, which means the test itself couldn't run (for example, invalid SQL).

:::

## Querying telemetry data

> Query the telemtry data to gain deeper insights into your dbt runs.

Reviewer: ```suggestion
Query the telemetry data to gain deeper insights into your dbt runs.
```

### JSONL examples

The following are some examples of querying the JSONL telemetry data. 

Watch for errors in real-time:
```bash
tail -f telemetry.jsonl | jq 'select(.severity_text == "ERROR")'
```

> Find all node execution results

Reviewer 2: oops, my bad, this example is not totally correct. can we remove it? I think 2 is enough for starters. 

We can probably discuss what is the best approach here as we can provide numerous ready made and useful “examples”, or we can let users go and figure them out using just 1-2. But I’d love to get this out ASAP and iterate later

```bash
cat telemetry.jsonl | jq 'select(.record_type == "SpanEnd" and .attributes.unique_id)'
```

List skipped nodes, reasons, and upstream details:
```bash

> cat telemetry.jsonl | jq 'select(.attributes.node_outcome == "NODE_OUTCOME_SKIPPED") | {node: .attributes.unique_id, reason: .attributes.node_skip_reason, upstream: .attributes.node_skip_upstream_detail.upstream_unique_id }'

Reviewer: is it uppercase or lowercase? the examples all display lowercase

Reviewer: if JSONL values are uppercase, we shoudl clarify that

Reviewer 2: yes, that is related to protobuf convention (protobuf define schema of the data). so in json we’ll have stuff like `NODE_OUTCOME_SKIPPED` although in other programmatic interface it will something like `NodeOutcome::Skipped` (rust) or `NodeOutcome.Skipped` (python). So there is a discrepancy between written language, programmatic APIs and json…

```python
duckdb.sql("""
    SELECT
        attributes.node_outcome,
        COUNT(*) as count
    FROM 'telemetry.parquet'
    WHERE attributes.node_outcome IS NOT NULL
    GROUP BY attributes.node_outcome
""").show()
```

## OpenTelemetry integration

<Constant name="fusion" />'s native OTLP support lets you send telemetry directly to any OpenTelemetry-compatible receiver, including Datadog, Jaeger, Google Cloud Trace, Grafana Tempo, and Honeycomb.

This enables:

- Integration with existing observability &mdash; No custom integrations needed.
- Custom alerts that trigger notifications on failures or slow builds.
- Correlate across systems that links dbt traces with downstream services.
- Centralized monitoring to view dbt alongside your other infrastructure.

<!-- :::tip Community example

 "I set up a basic OpenTelemetry collector that sends the dbt traces to Google Cloud Traces, which integrates with our devops observability, alarms, and on-call system. We now have a monitorable data ingestion stack.

The traces let us quickly find exactly which step failed, and the error log even provides a quick link into BigQuery showing us the exact generated query that failed."

::: -->

> ### Setting up OTLP export

Reviewer: can we make this super clear how the user has to set up env vars (e.g. that's what export is) and which env vars they can use? 

currently this section doesn't mention that OTLP export requires OpenTelemetry env vars or how to set it. also, are we missing another env var type? (see benoit's slack post below for the two env vars)

related, @b-per tried setting vars via a `.env` file and he didn't see the logs [see slack thread to follow](https://dbt-labs.slack.com/archives/C09PVME67U4/p1767869896251199)

Reviewer 2: I have PR that will load env vars from `.env` https://github.com/dbt-labs/fs/pull/6940 - BUT, it only applies to dbt env vars. OTLP env vars are canonical env vars dictated by open telemetry and read by open telemetry Rust SDK. So even if we will read `.env` files eventually, there wasn’t a plan (at least not yet) to read OTLP env vars…

one thing to consider is that whoever chooses to use OTLP will probably be familiar with them already, so it shouldn’t as obscure as dbt env vars for regular users… But i agree that the better we explain this the more love we’l get

Author: @mishamsk and I will address this in a follow up!

  "attributes": {
    "unique_id": "model.project.customers",
    "phase": "Run",
    "node_outcome": "success"
  }
}
```

| Field | Description |
| ----- | ----------- |
| `record_type` | `SpanStart`, `SpanEnd`, or `LogRecord` |
| `trace_id` | Unique identifier for the invocation (same data as `invocation_id` but in OTEL format). |
| `span_id` / `parent_span_id` | For reconstructing the span hierarchy. |
| `event_type` | Type identifier for filtering and parsing. |
| `attributes` | Event-specific data (schema varies by event type, but unlike OTEL conventions, it's strictly backed by a stable protobuf schema). |

## Schema stability

Unlike <Constant name="core" />'s structured logging, <Constant name="fusion" /> telemetry is backed by a public protobuf schema with strict compatibility guarantees:

- **Additive only** &mdash; New fields and event types may be added, but existing fields are never removed or changed.
- **Forward compatible** &mdash; Your integrations will continue to work as the schema evolves.

This makes <Constant name="fusion" /> telemetry a reliable foundation for production integrations, orchestrators, and long-term analytics pipelines.

> ## Official client library (coming soon)

Reviewer: can we avoid having things like 'coming soon' in the header to future proof ourselves. 
```suggestion
## Official client library (coming soon) {#official-client-library}
```

Reviewer 2: coming soon was from me, kinda self-inflicted commitment to make sure we deliver ;-)

Reviewer: huge big fan of committing! Just referring to removing the manual task of coming back to remove the 'coming soon' when we've done it. 🙏

> An official open-source client library is in development. Built in Rust for performance, it will be available as:

Reviewer: sugg active voice
```suggestion
dbt Labs is developing an official open-source client library. Built in Rust for performance, it will be available as:
```
